File: backend/internal/services/notification.go
package services

import (
	"database/sql"
	"fmt"
	"time"

	"github.com/google/uuid"
)

type NotificationService struct {
	db *sql.DB
}

func NewNotificationService(db *sql.DB) *NotificationService {
	return &NotificationService{db: db}
}

// CreateNotification creates a new notification for specified users
func (s *NotificationService) CreateNotification(userIDs []uuid.UUID, notifType, title, message string) error {
	if s.isQuietHours() {
		return nil
	}

	filteredUserIDs := s.filterUsersByPreferences(userIDs, notifType)
	if len(filteredUserIDs) == 0 {
		return nil
	}

	query := `INSERT INTO notifications (user_id, type, title, message, is_read, created_at) VALUES ($1, $2, $3, $4, false, NOW())`

	for _, userID := range filteredUserIDs {
		_, err := s.db.Exec(query, userID, notifType, title, message)
		if err != nil {
			return fmt.Errorf("failed to create notification for user %s: %v", userID, err)
		}
	}
	return nil
}

// CreateNotificationForRole creates notifications for all users with a specific role
func (s *NotificationService) CreateNotificationForRole(role, notifType, title, message string) error {
	query := `SELECT id FROM users WHERE role = $1 AND is_active = true`
	rows, err := s.db.Query(query, role)
	if err != nil {
		return fmt.Errorf("failed to fetch users by role: %v", err)
	}
	defer rows.Close()

	var userIDs []uuid.UUID
	for rows.Next() {
		var userID uuid.UUID
		if err := rows.Scan(&userID); err != nil {
			continue
		}
		userIDs = append(userIDs, userID)
	}

	if len(userIDs) == 0 {
		return nil
	}
	return s.CreateNotification(userIDs, notifType, title, message)
}

// isQuietHours checks if current time is within quiet hours
func (s *NotificationService) isQuietHours() bool {
	var quietHoursStart, quietHoursEnd string
	query := `SELECT COALESCE((SELECT setting_value FROM system_settings WHERE setting_key = 'quiet_hours_start'), '22:00'), COALESCE((SELECT setting_value FROM system_settings WHERE setting_key = 'quiet_hours_end'), '08:00')`
	err := s.db.QueryRow(query).Scan(&quietHoursStart, &quietHoursEnd)
	if err != nil {
		return false
	}

	now := time.Now()
	currentHour := now.Hour()

	startTime, err1 := time.Parse("15:04", quietHoursStart)
	endTime, err2 := time.Parse("15:04", quietHoursEnd)

	if err1 != nil || err2 != nil {
		return false
	}

	startHour := startTime.Hour()
	endHour := endTime.Hour()

	if startHour > endHour {
		return currentHour >= startHour || currentHour < endHour
	}
	return currentHour >= startHour && currentHour < endHour
}

// filterUsersByPreferences filters users based on their notification preferences
func (s *NotificationService) filterUsersByPreferences(userIDs []uuid.UUID, notifType string) []uuid.UUID {
	if len(userIDs) == 0 {
		return userIDs
	}

	prefColumn := "order_updates"
	switch notifType {
	case "order":
		prefColumn = "order_updates"
	case "inventory":
		prefColumn = "inventory_alerts"
	case "payment":
		prefColumn = "payment_notifications"
	case "system":
		prefColumn = "system_alerts"
	}

	query := fmt.Sprintf(`SELECT user_id FROM notification_preferences WHERE user_id = ANY($1) AND %s = true`, prefColumn)
	rows, err := s.db.Query(query, userIDs)
	if err != nil {
		return userIDs // Fail-open
	}
	defer rows.Close()

	var filteredIDs []uuid.UUID
	for rows.Next() {
		var userID uuid.UUID
		if err := rows.Scan(&userID); err != nil {
			continue
		}
		filteredIDs = append(filteredIDs, userID)
	}

	if len(filteredIDs) == 0 {
		return userIDs // Assume all want notifications if no preferences
	}
	return filteredIDs
}

// NotifyLowStock sends notification when stock is low
func (s *NotificationService) NotifyLowStock(productName string, currentStock, minStock int) error {
	title := "⚠️ Stok Menipis"
	message := fmt.Sprintf("%s tersisa %d (minimum: %d)", productName, currentStock, minStock)
	return s.CreateNotificationForRole("manager", "inventory", title, message)
}

// NotifyOrderCreated sends notification when a new order is created
func (s *NotificationService) NotifyOrderCreated(orderNumber, table string) error {
	title := fmt.Sprintf("Pesanan Baru #%s", orderNumber)
	message := fmt.Sprintf("Pesanan baru diterima untuk %s", table)
	return s.CreateNotificationForRole("kitchen", "order", title, message)
}

// NotifySystemAlert sends system-level notifications to admins
func (s *NotificationService) NotifySystemAlert(title, message string) error {
	return s.CreateNotificationForRole("admin", "system", title, message)
}

File: backend/src/services/notification.ts
import { pool } from '../db/connection.js';

// ── CreateNotification ───────────────────────────────────────────────────────
// Creates a notification for a specific user, respecting preferences and quiet hours.

export async function createNotification(
  userId: string,
  type: string,
  title: string,
  message: string,
): Promise<void> {
  try {
    // Check quiet hours
    if (await isQuietHours(userId)) return;

    // Check if user has this type enabled
    if (!(await isTypeEnabled(userId, type))) return;

    await pool.query(
      `INSERT INTO notifications (user_id, type, title, message)
       VALUES ($1, $2, $3, $4)`,
      [userId, type, title, message],
    );
  } catch (err) {
    console.error('Failed to create notification:', (err as Error).message);
  }
}

// ── CreateNotificationForRole ────────────────────────────────────────────────
// Creates a notification for all active users with a specific role.

export async function createNotificationForRole(
  role: string,
  type: string,
  title: string,
  message: string,
): Promise<void> {
  try {
    const usersRes = await pool.query(
      `SELECT id FROM users WHERE role = $1 AND is_active = true`,
      [role],
    );

    const filteredUsers = await filterUsersByPreferences(usersRes.rows, type);

    for (const user of filteredUsers) {
      await pool.query(
        `INSERT INTO notifications (user_id, type, title, message)
         VALUES ($1, $2, $3, $4)`,
        [user.id, type, title, message],
      );
    }
  } catch (err) {
    console.error('Failed to create notification for role:', (err as Error).message);
  }
}

// ── NotifyLowStock ───────────────────────────────────────────────────────────

export async function notifyLowStock(
  ingredientName: string,
  currentStock: number,
  minimumStock: number,
): Promise<void> {
  const message = `Low stock alert: ${ingredientName} is at ${currentStock} (minimum: ${minimumStock})`;

  // Notify admins and managers
  for (const role of ['admin', 'manager']) {
    await createNotificationForRole(role, 'low_stock', 'Low Stock Alert', message);
  }
}

// ── NotifyOrderCreated ───────────────────────────────────────────────────────

export async function notifyOrderCreated(
  orderNumber: string,
  orderType: string,
): Promise<void> {
  const message = `New ${orderType.replace('_', ' ')} order: ${orderNumber}`;

  // Notify kitchen staff
  await createNotificationForRole('kitchen', 'order_update', 'New Order', message);

  // Notify counters
  await createNotificationForRole('counter', 'order_update', 'New Order', message);
}

// ── NotifySystemAlert ────────────────────────────────────────────────────────

export async function notifySystemAlert(
  title: string,
  message: string,
): Promise<void> {
  await createNotificationForRole('admin', 'system_alert', title, message);
}

// ── Helper: isQuietHours ─────────────────────────────────────────────────────

async function isQuietHours(userId: string): Promise<boolean> {
  try {
    const res = await pool.query(
      `SELECT quiet_hours_start, quiet_hours_end
       FROM notification_preferences
       WHERE user_id = $1`,
      [userId],
    );

    if (res.rows.length === 0) return false;

    const { quiet_hours_start, quiet_hours_end } = res.rows[0];
    if (!quiet_hours_start || !quiet_hours_end) return false;

    // Get current time in WIB (UTC+7)
    const now = new Date();
    const wibTime = new Date(now.toLocaleString('en-US', { timeZone: 'Asia/Jakarta' }));
    const currentHour = wibTime.getHours();
    const currentMinutes = currentHour * 60 + wibTime.getMinutes();

    const startParts = quiet_hours_start.split(':');
    const endParts = quiet_hours_end.split(':');
    const startMinutes = parseInt(startParts[0], 10) * 60 + parseInt(startParts[1], 10);
    const endMinutes = parseInt(endParts[0], 10) * 60 + parseInt(endParts[1], 10);

    // Handle overnight quiet hours (e.g., 22:00 to 07:00)
    if (startMinutes > endMinutes) {
      return currentMinutes >= startMinutes || currentMinutes < endMinutes;
    }

    return currentMinutes >= startMinutes && currentMinutes < endMinutes;
  } catch {
    return false;
  }
}

// ── Helper: isTypeEnabled ────────────────────────────────────────────────────

async function isTypeEnabled(userId: string, type: string): Promise<boolean> {
  try {
    const res = await pool.query(
      `SELECT types_enabled FROM notification_preferences WHERE user_id = $1`,
      [userId],
    );

    if (res.rows.length === 0) return true; // Default: all enabled

    const typesEnabled = res.rows[0].types_enabled;
    if (!typesEnabled || !Array.isArray(typesEnabled)) return true;

    return typesEnabled.includes(type);
  } catch {
    return true; // Default: enabled on error
  }
}

// ── Helper: filterUsersByPreferences ─────────────────────────────────────────

async function filterUsersByPreferences(
  users: Array<{ id: string }>,
  type: string,
): Promise<Array<{ id: string }>> {
  const filtered: Array<{ id: string }> = [];

  for (const user of users) {
    // Check quiet hours
    if (await isQuietHours(user.id)) continue;

    // Check type preference
    if (!(await isTypeEnabled(user.id, type))) continue;

    filtered.push(user);
  }

  return filtered;
}
